http-clients/src/main/java/org/triplea/http/client/web/socket/GenericWebSocketClient.java

```diff
@@ -0,0 +1,94 @@
+package org.triplea.http.client.web.socket;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.gson.Gson;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.function.Consumer;
+import lombok.extern.java.Log;
+import org.triplea.java.concurrency.CompletableFutureUtils;
+
+/**
+ * A generic client class for creating websocket connections to a server. This client is based
+ * around a concept of uniform incoming and outgoing messages (eg: an envelope). This class handles
+ * the details of the underlying connection and provides methods to open/close the connection.
+ * Furthermore, this class provides a listener API to be notified of incoming message events, a send
+ * API, and it automatically converts incoming and outgoing messages to JSON string.
+ *
+ * @param <IncomingT> Message type we expect to receive from the server.
+ * @param <OutgoingT> Message type we send to the server.
+ */
+@Log
+public class GenericWebSocketClient<IncomingT, OutgoingT> implements WebSocketConnectionListener {
+  private static final Gson gson = new Gson();
+  private final ExecutorService threadPool = Executors.newFixedThreadPool(4);
+
+  private final WebSocketConnection client;
+  private final Class<IncomingT> incomingMessageType;
+  private final Consumer<IncomingT> messageListener;
+  private final Collection<Consumer<String>> connectionLostListeners = new ArrayList<>();
+
+  public GenericWebSocketClient(
+      final URI lobbyUri,
+      final Class<IncomingT> incomingMessageType,
+      final Consumer<IncomingT> messageListener) {
+    this(incomingMessageType, messageListener, new WebSocketConnection(lobbyUri));
+  }
+
+  @VisibleForTesting
+  GenericWebSocketClient(
+      final Class<IncomingT> incomingMessageType,
+      final Consumer<IncomingT> messageListener,
+      final WebSocketConnection webSocketClient) {
+    this.incomingMessageType = incomingMessageType;
+    this.messageListener = messageListener;
+    client = webSocketClient;
+    client.addListener(this);
+    client.connect();
+  }
+
+  /**
+   * Non-blocking send of a message to the server. Implementation note: data is sent as a JSON
+   * string, this method handles conversion of the parameter object to JSON.
+   *
+   * @param message The data object to send to the server.
+   */
+  public void send(final OutgoingT message) {
+    final String json = gson.toJson(message);
+    CompletableFutureUtils.logExceptionWhenComplete(
+        CompletableFuture.runAsync(() -> client.sendMessage(json), threadPool),
```

Comment:
Ok I dug into the code of the library and I'm pretty sure the send operation is asynchronous already. (Like I expected it to work from my experience with javascript)
So using a worker pool is probably not speeding things up by a noticeable amount.
Of course it might make sense to parallelize json serialization if the objects are large, but sending doesn't seem to be blocking at all.

Reply:
This `sendMessage` is calling:

```
  void sendMessage(final String message) {
    Preconditions.checkState(!closed);
    waitForConnection();
    if (client.isOpen()) {
      client.send(message);
    } else {
      log.severe("Failed to establish connection to server.");
    }
  }
```

Note the `waitForConnection`

Reply:
@RoiEXLab I extracted the socket connection establishing code to a new class 'WebSocketConnector', removed the wait loop and implemented the single threadpool blocking mechanism in the new connector class. I still have the 'threadpool' here though. The threadpool here will not guarantee ordering if there is any buffering, but I think that is okay. Errorprone really insists that the future object of 'sendMessage' be checked, which is a weak but actual reason for the threadpool to exist so we can have some code to do that logging. The conversion to JSON should be a pretty fast operation, all data is in memory, so even with a large payload we'd expect nano or micro-second timescales.

Let me know your thoughts on the updates.